### PR TITLE
roachtest: ensure privileges stay consistent after version upgrades

### LINKED ISSUE #65011: sql: add validating privileges after version upgrades
In 20.2, USAGE privilege was added incorrectly by displacing ZONECONFIG's bitfield value.

Not realizing Kind was a bitfield, I added USAGE above ZONECONFIG here:
https://github.com/cockroachdb/cockroach/pull/51622/files#diff-d4927b542ad295710e97b3e1adff3f107e31eb9dbd75a8adfbae4bdb4929a705R41

Thus after upgrading from 20.1 to 20.2, USAGE privilege would become ZONECONFIG. Very very fortunately, ZONECONFIG was only valid on tables/dbs whereas USAGE was not so we were able to tell which privilege the bit should be referring to and fix it.

Fix here:
https://github.com/cockroachdb/cockroach/pull/65010

We should introduce a test that ensures privileges stay consistent after upgrades.

Could do some sort of versionupgrade test.
https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/versionupgrade.go

Epic CRDB-2586

### PR DESCRIPTION
**Pull request by RichardJCai**
roachtest: ensure privileges stay consistent after version upgrades

Release note: None

Resolves https://github.com/cockroachdb/cockroach/issues/65011

### CODE REVIEW

**Review by rafiss — CHANGES_REQUESTED** [review]
this test is cool! hadn't looked at a version upgrade test before -- i like how you wrote this

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67375#-:-Me8WHKY-HO_Mr8NrMLj:bas5ubs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)

___
*[pkg/cmd/roachtest/privilege_version_upgrade.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67375#-Me8SPMRC2VjoAVBpDpn:-Me8SPMSDloKMjenR8yn:bmern67) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdbef6eac7336efd8e69927b08e5820874edf763/pkg/cmd/roachtest/privilege_version_upgrade.go#L30)):*
> ```Go
> 		Name:       "versionupgrade/privileges",
> 		Owner:      OwnerSQLExperience,
> 		MinVersion: "v20.1.0",
> ```

oh hm, i thought MinVersion got removed recently? i might be mistaken
___
*[pkg/cmd/roachtest/privilege_version_upgrade.go, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67375#-Me8TUAw6D6xP0F-Mx40:-Me8TUAw6D6xP0F-Mx41:b-9c6vdl) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdbef6eac7336efd8e69927b08e5820874edf763/pkg/cmd/roachtest/privilege_version_upgrade.go#L66)):*
> ```Go
> 			expectedGrants = append(expectedGrants, []string{"test", "testuser", privilege})
> 		}
> 		r.CheckQueryResults(t, `SHOW GRANTS ON DATABASE test`, expectedGrants)
> ```

is the order of the results stable? might be worth adding an extra `ORDER BY 1, 2, 3` to be sure. (same for the other SHOW GRANTS below
___
*[pkg/cmd/roachtest/privilege_version_upgrade.go, line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67375#-Me8UlUO5gfJABBsZATI:-Me8UlUO5gfJABBsZATJ:b-kcjo6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdbef6eac7336efd8e69927b08e5820874edf763/pkg/cmd/roachtest/privilege_version_upgrade.go#L206)):*
> ```Go
> 	// An empty string will lead to the cockroach binary specified by flag
> 	// `cockroach` to be used.
> 	const mainVersion = ""
> ```

nit: would this be better named `currentVersion` or something? also, why don't we want to use `buildVersion` here?
___
*[pkg/cmd/roachtest/privilege_version_upgrade.go, line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67375#-Me8UU589vSU032ksP8c:-Me8UU589vSU032ksP8d:b-fdo42s) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdbef6eac7336efd8e69927b08e5820874edf763/pkg/cmd/roachtest/privilege_version_upgrade.go#L221)):*
> ```Go
> 		steps = append(
> 			steps,
> 			createUserStep(1),
> ```

nit: since loadNode is always 1, could it be defined as a const rather than passed as a parameter?
___
*[pkg/cmd/roachtest/privilege_version_upgrade.go, line 273 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67375#-Me8Vs4t75LpAYY45P_Q:-Me8Vs4t75LpAYY45P_R:b-vrakth) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdbef6eac7336efd8e69927b08e5820874edf763/pkg/cmd/roachtest/privilege_version_upgrade.go#L273)):*
> ```Go
> 			continue
> 		}
> 		if i%2 == 0 {
> ```

i thought for a second if this would be a good place to randomize, but i'm not sure. of course the downside is that it could get annoying to reproduce, but on the pro side, the test will always log the expected privileges it it fails and we might get more coverage... idk if it's worth it though

**Review by RichardJCai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67375)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/cmd/roachtest/privilege_version_upgrade.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67375#-Me8SPMRC2VjoAVBpDpn:-MeAzcDlEwA15CkJRMbf:bfyw65m) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdbef6eac7336efd8e69927b08e5820874edf763/pkg/cmd/roachtest/privilege_version_upgrade.go#L30)):*
> ```
> nVer
> ```
Yeah you're right looks like its gone
___
*[pkg/cmd/roachtest/privilege_version_upgrade.go, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67375#-Me8TUAw6D6xP0F-Mx40:-MeAzd_C8LbGRu_mB6XD:b-30elu) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdbef6eac7336efd8e69927b08e5820874edf763/pkg/cmd/roachtest/privilege_version_upgrade.go#L66)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is the order of the results stable? might be worth adding an extra `ORDER BY 1, 2, 3` to be sure. (same for the other SHOW GRANTS below
</blockquote></details>

The underlying SHOW GRANTS query actually has an `ORDER BY 1,2,3`
___
*[pkg/cmd/roachtest/privilege_version_upgrade.go, line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67375#-Me8UlUO5gfJABBsZATI:-MeAzhVu5qcr_fMUs20U:b-w5sar6) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdbef6eac7336efd8e69927b08e5820874edf763/pkg/cmd/roachtest/privilege_version_upgrade.go#L206)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: would this be better named `currentVersion` or something? also, why don't we want to use `buildVersion` here?
</blockquote></details>

Yeah I think `currentVersion` makes sense.  buildVersion probably also makes sense
___
*[pkg/cmd/roachtest/privilege_version_upgrade.go, line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67375#-Me8UU589vSU032ksP8c:-MeAzu-DDNyUj6hiT29V:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdbef6eac7336efd8e69927b08e5820874edf763/pkg/cmd/roachtest/privilege_version_upgrade.go#L221)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: since loadNode is always 1, could it be defined as a const rather than passed as a parameter?
</blockquote></details>

Done
___
*[pkg/cmd/roachtest/privilege_version_upgrade.go, line 273 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67375#-Me8Vs4t75LpAYY45P_Q:-MeAzunj0BZ8nw5m15GH:b-bbn4by) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdbef6eac7336efd8e69927b08e5820874edf763/pkg/cmd/roachtest/privilege_version_upgrade.go#L273)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i thought for a second if this would be a good place to randomize, but i'm not sure. of course the downside is that it could get annoying to reproduce, but on the pro side, the test will always log the expected privileges it it fails and we might get more coverage... idk if it's worth it though
</blockquote></details>

Yeah imo I don't think it's worth it to randomize.
I don't think randomizing will do much since we're just trying to make sure each privilege stays consistent after upgrading. Testing different combinations won't really expand coverage.

**Review by RichardJCai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67375)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/cmd/roachtest/privilege_version_upgrade.go, line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67375#-Me8UlUO5gfJABBsZATI:-MeBb-Bf7ijyNHSJeXqv:b-hia7t8) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdbef6eac7336efd8e69927b08e5820874edf763/pkg/cmd/roachtest/privilege_version_upgrade.go#L206)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Yeah I think `currentVersion` makes sense.  buildVersion probably also makes sense
</blockquote></details>

Nvm using buildVersion errors out, seems like it doesn't find the right binary name or the url doesn't match the buildVersion exactly.

```
		Wraps: (3) /Users/rcai/go/src/github.com/cockroachdb/cockroach/bin/roachprod stage richardc-1625856233-01-n3cpu4:3 release vv21.2.0-alpha.00000000 --dir='vv21.2.0-alpha.00000000' returned
		  | stderr:
		  | curl: (22) The requested URL returned error: 404 Not Found
		  |
		  | gzip: stdin: unexpected end of file
		  | tar: Child returned status 1
		  | tar: Error is not recoverable: exiting now
		  | Error: COMMAND_PROBLEM: exit status 2
		  | (1) COMMAND_PROBLEM
		  | Wraps: (2) Node 3. Command with error:
		  |   | ```
		  |   |
		  |   | tmpdir="$(mktemp -d /tmp/cockroach-release.XXX)" && \
		  |   | dir='vv21.2.0-alpha.00000000' && \
		  |   | curl -f -s -S -o- https://s3.amazonaws.com/binaries.cockroachdb.com/cockroach-vv21.2.0-alpha.00000000.linux-amd64.tgz | tar xfz - -C "${tmpdir}" --strip-components 1 && \
		  |   | mv ${tmpdir}/cockroach ${dir}/cockroach && \
		  |   | mkdir -p ${dir}/lib && \
		  |   | if [ -d ${tmpdir}/lib ]; then mv ${tmpdir}/lib/* ${dir}/lib; fi && \
		  |   | chmod 755 ${dir}/cockroach
		  |   |
		  |   | ```
		  | Wraps: (3) exit status 2
		  | Error types: (1) errors.Cmd (2) *hintdetail.withDetail (3) *exec.ExitError
		  |
		  | stdout:
		  | Resolved release url for cockroach version vv21.2.0-alpha.00000000: https://s3.amazonaws.com/binaries.cockroachdb.com/cockroach-vv21.2.0-alpha.00000000.linux-amd64.tgz
```

**Comment by rafiss** [comment]
Seems like some changes didn't get pushed

**Review by RichardJCai — COMMENTED** [review]
Oops yeah I think I forgot to add the file or something

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67375#-:-MeBj81O2vvCNjpRVdx3:burbncq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

**Review by rafiss — APPROVED** [review]
looks good!

> Reviewed 2 of 3 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67375#-:-MeVpOD-2sTNRNTNkS2n:bbo24j7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

**Comment by RichardJCai** [comment]
Thanks!
bors r=rafiss

File: pkg/cmd/roachtest/BUILD.bazel [hunk 1]
```diff
@@ -53,14 +53,12 @@ go_binary(
...
 go_test(
     name = "roachtest_test",
-    size = "small",
     srcs = [
         "z_cluster_test.go",
         "z_test_registry_test.go",
         "z_test_test.go",
     ],
     embed = [":roachtest_lib"],
-    tags = ["broken_in_bazel"],
```
Comment (rickystewart) [inline pkg/cmd/roachtest/BUILD.bazel:63]: Careful -- the removal of this tag started breaking Bazel CI runs. :)

Comment (RichardJCai) [reply pkg/cmd/roachtest/BUILD.bazel:63]: Interesting, thanks for fixing. Odd enough I think the tag got removed by `make bazel-generate`. I'll keep this in mind for the future.

**Pull request merged**